Contributing.md

```diff
@@ -1,30 +1,472 @@
+at all (since PRs [do one job]). 
+
+In some circumstances, green-field code can break this rule. But then you really MUST make reviewers' lives as easy 
+as possible by offering multiple commits, reams of documentation, git commit messages and helpful tests.
+
+The 400 line limit represents about an hour of solid code review time. Research indicates that spending more than 
+this in a single session leads to significantly more bugs scuttling through the gate.
+
+If your PR is long, you can expect a reviewer to label it `CR-too_long` and ask you to break it up.
+
+If you really, really, can't break it up, then do some [git commit farming] to break the PR up into PR-sized commits.
+
+### Use git messages liberally
+
+Before submitting your PR for the first time, use `git rebase` as described in [git commit farming] to edit and 
+clean up your commit messages. The messages should supplement the documentation in the code, not just repeat it. 
+This contains the meta-information; _why_ you are making a change, rather than _what_ it does.
+
+For example, a lousy commit message is:
+
+```text
+Refactor signature check
+```
+
+We can see that you've changed the signature check code! But we don't know why. 
+
+A better message is
+
+```text
+The signature check was checking each signature in the block independently. So for n signatures, the time-scaled by
+O(n).
+We can optimise this check by making use of an aggregated signature check, as described by 
+[Alice and Bob](https://y.at/yat.coin). This was codified and included in tari_crypto PR #1234.
+Changing the signature check to use this improved overall block validation times by 65% (see benchmarks in the
+next commit). 
+
+All tests still pass but could improve coverage by adding additional tests with large numbers of 
+signatures (TODO - not in this PR)
+```
+
+You may find that a reviewer tags your PR with `CR-insufficient_context`. He is asking you to add 
+additional context for the change. This can be submitted by tidying up and expanding your git commit messages, the 
+PR description, or both.
+
+The commit messages are combined to form the Pull Request's narrative when the PR is merged.
+Ideally, if someone reads the `git log` for your PR, they'll have a clear picture of why and how the changes were 
+introduced.
+
+## Code reviews
+
+Post mainnet, PRs require 3 or more positive reviews from previous contributors, including at least one positive review 
+from a contributor with write access to the repo.
+
+When reviewing PRs, here are some guidelines and suggestions to help maintain the safety and quality of code:
+
+* PULL the code and review and test it locally on your machine.
+* Do not spend more than 1 hour reviewing a PR at a time. 
+* Do not review more than 400 LOC in an hour.
+
+If either of these guidelines cannot be met, you may label the PR `CR-too_long` and politely ask the contributor to 
+revise their PR. Otherwise, take breaks and review the PR over multiple sessions or days.
```

Comment:
I think the 1 hour guideline here is not practical or clear enough. For example, setting up a base node, wallet, miner, etc., compiling release code, syncing to the specific testnet, doing some system level tests, etc. will take way more time than this.

Maybe 1 hour just means reading, evaluating and browsing the code?

Reply:
I'm leaving it to the discretion of the reviewer.